src/blob_store/store.rs

```diff
@@ -0,0 +1,493 @@
+use crate::blob_store::recordfile::RecordFile;
+use crate::blob_store::slot::{SlotData, SlotIO};
+use crate::blob_store::store_impl as simpl;
+use crate::blob_store::{Result, SlotMeta, StoreConfig, StoreError};
+use crate::entry::Entry;
+use crate::packet::{Blob, BLOB_HEADER_SIZE};
+
+use byteorder::{BigEndian, ByteOrder};
+
+use std::borrow::Borrow;
+use std::collections::BTreeMap;
+use std::fmt;
+use std::fs::{File, OpenOptions};
+use std::io::prelude::*;
+use std::ops::Range;
+use std::path::{Path, PathBuf};
+use std::result::Result as StdRes;
+
+pub const DEFAULT_SLOT_CACHE_SIZE: usize = 10;
+pub const DATA_FILE_BUF_SIZE: usize = 64 * 1024;
+pub const INDEX_RECORD_SIZE: u64 = 3 * 8;
+
+const SLOT_REC_NAME: &str = "slots";
+
+#[derive(Debug)]
+pub struct Store {
+    root: PathBuf,
+    config: StoreConfig,
+    cache: SlotCache,
+    slots_rec: RecordFile<u64>,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Key(pub u64, pub u64);
+
+#[derive(Debug)]
+pub struct SlotCache {
+    max_size: usize,
+    map: BTreeMap<u64, SlotIO>,
+}
+
+pub trait Storable {
+    type ToErr: fmt::Display;
+
+    fn to_data(&self) -> StdRes<Vec<u8>, Self::ToErr>;
+}
+
+pub trait StorableNoCopy: Storable {
+    fn as_data(&self) -> StdRes<&[u8], Self::ToErr>;
+}
+
+pub trait Retrievable {
+    type FromErr: fmt::Display;
+    type Output;
+
+    fn from_data(data: &[u8]) -> StdRes<Self::Output, Self::FromErr>;
+}
+
+pub trait Named {
+    const COLUMN: &'static str;
+}
+
+pub trait Column: Named + Storable + Retrievable {}
+
+pub trait ColumnSingle: Named + Storable + Retrievable {}
+
+pub trait ColumnNoCopy: Column + StorableNoCopy {}
+
+impl Store {
+    pub fn open<P: AsRef<Path>>(path: &P) -> Result<Store> {
+        Store::with_config(path, StoreConfig::default())
+    }
+
+    pub fn with_config<P: AsRef<Path>>(path: &P, config: StoreConfig) -> Result<Store> {
+        Ok(Store {
+            root: PathBuf::from(path.as_ref()),
+            config,
+            cache: SlotCache::default(),
+            slots_rec: RecordFile::open(path.as_ref().join(SLOT_REC_NAME))?,
+        })
+    }
+
+    pub fn config(&self) -> &StoreConfig {
+        &self.config
+    }
+
+    #[inline]
+    pub fn put<T>(&mut self, key: Key, obj: T) -> Result<()>
+    where
+        T: Column,
+    {
+        self.put_dyn(T::COLUMN, key, obj)
+    }
+
+    #[inline]
+    pub fn put_dyn<T>(&mut self, column: &str, key: Key, obj: T) -> Result<()>
```

Comment:
These methods (gets/puts) are mostly called with the same type of T. Any good reason to use generics here?

Reply:
The intention was just to make things generic so that I wasn't duplicating code. The traits are definitely excessive I think. They'll be replaced with methods that just accept any `T: Serialize` for put and  returns bytes or any `T: Deserialize` for gets.